feat(ssh): auto generate additional ssh keys

TheFox0x7 opened the pull request:
adds capabilities for gitea to generate ecdsa and ed25519 keys by default
adds cli for built-in ssh key generation helpers


closes: https://github.com/go-gitea/gitea/issues/33783

TheFox0x7 committed:
add test for keygen

TheFox0x7 committed:
swap to PKCS8 form

this allows storing different keys

TheFox0x7 committed:
prepare test for more keys

TheFox0x7 committed:
add ecdsa and ed25519

TheFox0x7 committed:
fix formatting

TheFox0x7 committed:
add copyright

TheFox0x7 committed:
extend settings

rework keygen to guess the keysize from extension

TheFox0x7 committed:
remove failure possibilty

TheFox0x7 committed:
add keygen command

make key generation happen in generation module

TheFox0x7 committed:
reuse generate module

TheFox0x7 committed:
fix bad defaults

TheFox0x7 committed:
fix checks

TheFox0x7 committed:
remove blankline

wxiaoguang commented on cmd/generate.go:
```diff
@@ -98,3 +112,32 @@ func runGenerateSecretKey(c *cli.Context) error {
+	if err != nil {
+		return err
+	}
+	f, err := os.OpenFile(file, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0o600)
```
Should it avoid overwriting existing file?

```
/tmp$ ssh-keygen -t ecdsa -f a
Generating public/private ecdsa key pair.
a already exists.
Overwrite (y/n)?
```

wxiaoguang commented on modules/ssh/ssh.go:
```diff
@@ -400,12 +406,21 @@ func Listen(host string, port int, ciphers, keyExchanges, macs []string) {
 func GenKeyPair(keyPath string) error {
-	privateKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	bits := 4096
+	keytype := filepath.Ext(keyPath)
```
It doesn't seem right to guess the content by file extension. The key itself has the correct information.

```
$ ssh-keygen -t ecdsa -f a
...
$ ssh-keygen -lf a
256 SHA256:NS1YmfwZDYoEgtS3Ne0wRuDvGFFWyQfc1QB8ZT2SGeA xiaoguang@Xiaoguangs-MacBook-Pro.local (ECDSA)
```

wxiaoguang commented on modules/ssh/ssh.go:
```diff
@@ -367,18 +373,19 @@ func Listen(host string, port int, ciphers, keyExchanges, macs []string) {
-		err := GenKeyPair(setting.SSH.ServerHostKeys[0])
-		if err != nil {
-			log.Fatal("Failed to generate private key: %v", err)
+		for i := range 3 {
```
Why `3`? Why generates all files?

TheFox0x7 committed:
add overwrite protection and prompt

TheFox0x7 committed:
add keyinit to better mimick behavior of ssh-keygen

TheFox0x7 committed:
Merge branch 'main' into generate-additional-ssh-keys

TheFox0x7 committed:
fix integration test call

TheFox0x7 committed:
Merge branch 'main' into generate-additional-ssh-keys

TheFox0x7 committed:
fix formatting

TheFox0x7 committed:
fix leftover from urfavev2

TheFox0x7 committed:
fix argument type

Assisted-By: mistral-vibe:mistral-medium3.5

TheFox0x7 committed:
Merge branch 'main' into generate-additional-ssh-keys

bircni committed:
Merge branch 'main' into generate-additional-ssh-keys

TheFox0x7 committed:
Merge branch 'main' into generate-additional-ssh-keys

lunny reviewed (approved):
It's better to use KeyType instead of string. Otherwise LGTM. https://github.com/go-gitea/gitea/pull/33974/files#r3350927882

TheFox0x7 committed:
use KeyType in ssh module

doesn't pass it to generate as it would make circluar dependancy

bircni commented:
LGTM for me too but @wxiaoguang s Feedback would be nice

bircni committed:
Merge branch 'main' into generate-additional-ssh-keys

wxiaoguang committed:
refactor

wxiaoguang committed:
refactor

wxiaoguang committed:
fine tune

wxiaoguang reviewed (commented):
Refactored some code. Please double check whether the changes are right.

wxiaoguang commented on modules/ssh/ssh.go:
```diff
@@ -395,43 +389,45 @@ func Listen(host string, port int, ciphers, keyExchanges, macs []string) {
+		keyPath := filepath.Join(path, "gitea."+string(keyType))
+		_, errStatPriv := os.Stat(keyPath)
+		_, errStatPub := os.Stat(keyPath + ".pub")
+		if errStatPriv != nil || errStatPub != nil {
```
@TheFox0x7 

does it really need to check whether ".pub" file exists?

* Old logic: only private key is needed
* New logic: if "pub" file doesn't exist, it regenerates the private key, and it will break the remembered fingerprint on client side.

TheFox0x7 committed:
regenerate public key if private is avalible

surface host key generation in cli command so manual setup of keys is
possible on old versions of gitea

TheFox0x7 commented:
Thanks for the fixes, I didn't see anything wrong. One last thing I added is `-A` flag to the cli which runs the full host key generator just so old instances can get the new keys with one command since the generator won't run if there's a valid key

TheFox0x7 committed:
Merge branch 'main' into generate-additional-ssh-keys

TheFox0x7 committed:
drop public key regeneration

TheFox0x7 committed:
redo cli commands

GiteaBot committed:
Merge branch 'main' into generate-additional-ssh-keys

GiteaBot committed:
Merge branch 'main' into generate-additional-ssh-keys

GiteaBot committed:
Merge branch 'main' into generate-additional-ssh-keys

GiteaBot committed:
Merge branch 'main' into generate-additional-ssh-keys

GiteaBot committed:
Merge branch 'main' into generate-additional-ssh-keys

GiteaBot committed:
Merge branch 'main' into generate-additional-ssh-keys

GiteaBot committed:
Merge branch 'main' into generate-additional-ssh-keys

GiteaBot committed:
Merge branch 'main' into generate-additional-ssh-keys
